src/upload-button.tsx

```diff
@@ -0,0 +1,356 @@
+            const inputFile = event.target.files[i];
+
+            if (resolution && applyToAll && resolution.skip)
+                break;
+
+            if (resolution && applyToAll && resolution.replace)
+                toUploadFiles.push(inputFile);
+            else {
+                let uploadFile = true;
+                for (const file of files) {
+                    if (file.name === inputFile.name) {
+                        try {
+                            resolution = await Dialogs.show(
+                                <FileConflictDialog
+                                  path={path}
+                                  file={file}
+                                  uploadFile={inputFile}
+                                  showApplyAll={event.target.files.length > 1}
+                                />
+                            );
+                        } catch (exc) {
+                            resetInput();
+                            return;
+                        }
+
+                        if (resolution && resolution.skip)
+                            uploadFile = false;
+
+                        if (resolution.applyToAll)
+                            applyToAll = true;
+
+                        break;
+                    }
+                }
+
+                if (uploadFile)
+                    toUploadFiles.push(inputFile);
+            }
+        }
+
+        if (toUploadFiles.length === 0) {
+            resetInput();
+            return;
+        }
+
+        window.addEventListener("beforeunload", beforeUnloadHandler);
+
+        const cancelledUploads = [];
+        await Promise.allSettled(toUploadFiles.map(async (file: File) => {
+            const tmp_path = path.slice();
+            tmp_path.push(file.name);
+            const destination = tmp_path.join('/');
+            const abort = new AbortController();
+
+            setUploadedFiles(oldFiles => {
+                return {
+                    [file.name]: { file, progress: 0, cancel: () => abort.abort() },
+                    ...oldFiles,
+                };
+            });
+
+            try {
+                await upload(destination, file, (progress) => {
+                    const now = performance.now();
+                    if (now < next_progress)
+                        return;
+                    next_progress = now + 200; // only rerender every 200ms
+                    setUploadedFiles(oldFiles => {
+                        const oldFile = oldFiles[file.name];
+                        return {
+                            ...oldFiles,
+                            [file.name]: { ...oldFile, progress },
+                        };
+                    });
+                }, abort.signal);
+                // TODO: pass { superuser: try } depending on directory owner
+            } catch (exc) {
+                cockpit.assert(exc instanceof Error, "Unknown exception type");
```

Comment:
Can we be *really* sure of that? There can't be browser-triggered errors/aborts/network errors and what have you? I'd feel better if "any other exception" was just handled in the `else` branch, and conditionalize the `.toString()` instead.

Reply:
We're not really exposed to the network errors that the browser might throw in any direct way here.  Those are behind several layers of `cockpit.js` API (and in another frame)...

From my reckoning, it should really only be possible to have the error from the abort and the errors that we throw in response to a close message with a problem field.

All the same, who knows, right?  Maybe some day we get some weird exception from calling `performance.now()` or something?  I think I agree that I'd prefer to see this handled a bit more elegantly, but not a blocker from me.